plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClientModule.java

```diff
@@ -42,13 +49,53 @@ public void configure(Binder binder)
         binder.bind(JdbcClient.class).annotatedWith(ForBaseJdbc.class).to(ClickHouseClient.class).in(Scopes.SINGLETON);
         bindTablePropertiesProvider(binder, ClickHouseTableProperties.class);
         binder.install(new DecimalModule());
+        binder.bind(ConnectionFactory.class).annotatedWith(ForBaseJdbc.class).toProvider(ConnectionFactoryProvider.class).in(Scopes.SINGLETON);
     }
 
-    @Provides
-    @Singleton
-    @ForBaseJdbc
-    public static ConnectionFactory createConnectionFactory(BaseJdbcConfig config, CredentialProvider credentialProvider)
+    private static class ConnectionFactoryProvider
+            implements Provider<ConnectionFactory>
     {
-        return new ClickHouseConnectionFactory(new DriverConnectionFactory(new ClickHouseDriver(), config, credentialProvider));
+        private final ConnectionFactory connectionFactory;
+
+        @Inject
+        public ConnectionFactoryProvider(ClickHouseConfig clickHouseConfig, BaseJdbcConfig baseJdbcConfig, CredentialProvider credentialProvider)
+        {
+            connectionFactory = new ClickHouseConnectionFactory(
+                    new DriverConnectionFactory(createDriver(clickHouseConfig.isUseDeprecatedDriver()), baseJdbcConfig, credentialProvider));
+        }
+
+        @Override
+        public ConnectionFactory get()
+        {
+            return connectionFactory;
+        }
+
+        Driver createDriver(boolean useDeprecatedDriver)
+        {
+            String driverClass = getDriverClassName(useDeprecatedDriver);
+            try {
+                return (Driver) Class.forName(driverClass, true, getClassLoader()).getDeclaredConstructor().newInstance();
+            }
+            catch (ReflectiveOperationException e) {
+                throw new RuntimeException(format("Error creating an instance of %s", driverClass), e);
+            }
+        }
+
+        String getDriverClassName(boolean useDeprecatedDriver)
+        {
+            if (useDeprecatedDriver) {
+                return CLICKHOUSE_DEPRECATED_DRIVER_CLASS_NAME;
+            }
+            return CLICKHOUSE_LATEST_DRIVER_CLASS_NAME;
+        }
+
+        ClassLoader getClassLoader()
+        {
+            ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+            if (classLoader == null) {
+                classLoader = ClickHouseConfig.class.getClassLoader();
+            }
+            return classLoader;
+        }
```

Comment:
Any reason you avoided calling each constructor? 
```java
    private static Driver getDriver(boolean useDeprecatedDriver)
    {
        if (useDeprecatedDriver) {
            return new ru.yandex.clickhouse.ClickHouseDriver();
        }
        return new com.clickhouse.jdbc.ClickHouseDriver();
    }
```

Reply:
In `com.clickhouse:clickhouse-jdbc:0.3.2*` both old and new drivers exist, but after 0.4.x the old driver will be removed, in order to be able to upgrade to future 0.4.x without damage, so here I think it is more friendly to use reflection (no need to submit a new patch except for version changes).

As you suggested, upgrading to 0.4.x can also be done by submitting a new patch, so doing it via an explicit constructor is also desirable.

I think both ways are possible and would like to hear your opinion.